docs/maintainer/infrastructure.md

```diff
@@ -509,15 +509,172 @@ This was sponsored by Prefix.dev. Below are the relevant pull requests for the s
+and/or systems (e.g., C, Fortran, C++, CUDA, etc.). These are used
 in all of our CI builds to build essentially all artefacts published by conda-forge.
 
 This compiler infrastructure has a critical role beyond building everything, which
 is to ensure that packages stay compatible with each other. This is due to how compiled
 packages have a so-called [Application Binary Interface](../glossary.md#abi)
 (ABI), and how changes in the compiler infrastructure may break this ABI, leading
 to crashes, miscalculations, etc. Generally speaking, using a consistent compiler
-version greatly reduces the risk of ABI breaks.
+setup greatly reduces the risk of ABI breaks.
+
+### Using compilers in feedstocks
+
+There are two kinds of compiler-related packages in Conda-forge: implementation packages
+that install the compiler itself, and _activation_ packages that install scripts
+that set the build environment to use the respective compiler by default. These
+scripts set up a number of standard environment variables such as `CC`, and perform
+setup actions for the common build systems such as CMake and Meson.
+
+Recipes provide a `compiler` macro that is used to generate the correct dependency
+on a compiler for the language specified. Compilers are always added to `build`
+dependencies. For example, to depend on a C and C++ compilers, one would specify:
+
+```yaml
+requirements:
+  build:
+    - {{ compiler('c') }}
+    - {{ compiler('cxx') }}
+    - {{ stdlib('c') }}
+```
+
+The possible parameter values are listed in the [compilers supplied by conda-forge
+](#compilers-supplied-by-conda-forge) section.
+
+Almost always, in addition to a compiler of any programming language, you will
+also need to depend on the C standard library via the `stdlib('c')` macro.
+Note that currently this macro is only used to refer to the C standard library
+and in the vast majority of cases all programming languages use that library,
+therefore the macro is always used with the `'c'` parameter. For example,
+a regular Rust package would use:
+
+```yaml
+requirements:
+  build:
+    - {{ compiler('rust') }}
+    - {{ stdlib('c') }}
+```
+
+### Using non-default compilers in feedstocks
+
+Our default compiler stack is made up very differently on each platform; each platform
+has its own default compiler, with its own set of feedstocks that provide them.
+However, it is often possible to use a different compiler if necessary. For example,
+to use Clang 21 as the C and C++ compiler _on all platforms_ (and not just macOS), one would additionally specify
+in `recipe/conda_build_config.yaml`:
+
+```yaml
+c_compiler:
+  - clang
+c_compiler_version:
+  - 21
+cxx_compiler:
+  - clangxx
+cxx_compiler_version:
+  - 21
+```
+
+Alternatively, the specific compiler name can be used as an argument to
+the `{{ compiler(...) }}` macro, e.g.:
+
+```yaml
+requirements:
+  build:
+    - {{ compiler('clang') }}
+    - {{ stdlib('c') }}
+```
```

Comment:
TIL! Nice. Is this officially supported or more like a nice byproduct of how `compiler()` works (by adding the `_{build_platform}` subdir)?

Reply:
nvm, found my answer in https://github.com/conda-forge/conda-forge.github.io/pull/2596#discussion_r2379335994

Reply:
It's supported. If `clang_compiler` is set in cbc, it's used. Otherwise `clang` is used.

Reply:
> nvm, found my answer in [#2596 (comment)](https://github.com/conda-forge/conda-forge.github.io/pull/2596#discussion_r2379335994)

I cannot find the source of that statement in that thread. If you're basing this on my comment about `clang` vs. `clang-cl`, then I was talking about the _values_ of `c_compiler:` in CBC, _not_ the keys used in the `compiler()` macro.

TBH, I think `{{ compiler('clang') }}` is highly confusing and we should not be supporting it _at all_ in official docs. It works by accident because of the way that `{{ compiler("foo") }}` will work if you set `foo_compiler:` in CBC and the package-name constructed with `<value_of_foo_compiler>_{{ target_platform }}` happens to be legit.

Reply:
@h-vetinari, I've added it per https://github.com/conda-forge/conda-forge.github.io/pull/2596#discussion_r2379342036.

Reply:
Well, I'm fine with removing it or listing it as an anti-pattern (and submitting fixes to feedstocks doing that), but I'd like @isuruf to confirm, since (unless I misunderstood) he requested that.

Reply:
I've also applied the suggestions I've proposed above. I'll add `compiler('clang')` usage once/if the pinning PR is merged.